File: examples/javascript/data-types/data-types.js
// 第 09 章 · 数据类型 — JavaScript 示例
// 运行：node data-types.js

// 1. 只有一个 number 类型（双精度浮点），没有整数类型
let score = 92;
console.log("score 是整数值吗:", Number.isInteger(score), "| typeof:", typeof score);

// 2. 浮点误差（IEEE 754，所有语言共有）
console.log("0.1 + 0.2 =", 0.1 + 0.2);
console.log("等于 0.3 吗:", 0.1 + 0.2 === 0.3);
console.log("用容差比较:", Math.abs((0.1 + 0.2) - 0.3) < 1e-9);

// 3. 安全整数上限：大 ID 会丢精度
console.log("MAX_SAFE_INTEGER =", Number.MAX_SAFE_INTEGER);
console.log("超出后:", 9007199254740993, "← 末位被改了");
console.log("用 BigInt 才精确:", 9007199254740993n + 2n);

// 4. 字符串长度数的是 UTF-16 码元
const wave = "👋";
console.log("'👋'.length =", wave.length, "| 按字符数 =", [...wave].length);

// 5. 两种空值
let a;
console.log("undefined:", a, "| null:", null);

File: examples/python/data-types/data-types.py
# 第 09 章 · 数据类型 — Python 示例
# 运行：python3 data-types.py

from decimal import Decimal

# 1. int 是任意精度，永不溢出
score = 92
big = 2 ** 200
print("2**200 有", len(str(big)), "位数字，且完全精确")

# 2. float 就是 C 的 double，同样有 IEEE 754 误差
print("0.1 + 0.2 =", 0.1 + 0.2)
print("等于 0.3 吗:", 0.1 + 0.2 == 0.3)

# 3. 需要精确小数用 Decimal（必须从字符串构造）
print("Decimal 精确:", Decimal("0.1") + Decimal("0.2"),
      "| 等于 0.3 吗:", Decimal("0.1") + Decimal("0.2") == Decimal("0.3"))
print("从浮点构造是错的:", str(Decimal(0.1))[:30], "...")

# 4. 字符串长度数的是码点
wave = "👋"
print("len('👋') =", len(wave), "| UTF-8 字节数 =", len(wave.encode("utf-8")))

# 5. bool 是 int 的子类
print("True + True =", True + True, "| bool 是 int 吗:", isinstance(True, int))

# 6. 空值只有一个
print("空值:", None)
